src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -2607,6 +2595,29 @@ private void doCancelPublishExpirationIfNecessary()
                     publishExpiresId = NO_CANCEL_ID;
                 }
             }
+
+            private void updatePublishCapabilityIfNecessary(
+                long traceId,
+                long authorization)
+            {
+                if (!hasPublishCapability(capabilities))
+                {
+                    this.capabilities |= PUBLISH_ONLY.value();
+                    doApplicationFlush(traceId, authorization, 0);
+                }
+            }
+
+            private boolean hasPublishCapability(
+                int capabilities)
+            {
+                return (capabilities & PUBLISH_ONLY.value()) != 0;
+            }
+
+            private boolean hasSubscribeCapability(
+                int capabilities)
+            {
+                return (capabilities & SUBSCRIBE_ONLY.value()) != 0;
+            }
```

Comment:
Please move `hasPublishCapability` and `hasSubscribeCapability` to outer factory class, unless they are required to be here, but it doesn't seem so.